### PR TITLE
Add time_units to python API

### PR DESCRIPTION
**Pull request by benjeffery**
Part of #1644 

Needs docs and change log. 
Next PR is to add this to the tree sequence in both C and Python.

### CODE REVIEW

File: python/lwt_interface/tskit_lwt_interface.h [hunk 1]
```diff
@@ -1334,6 +1335,27 @@ parse_table_collection_dict(tsk_table_collection_t *tables, PyObject *tables_dic
...
+    value = get_table_dict_value(tables_dict, "time_units", false);
+    if (value == NULL) {
+        goto out;
+    }
+    if (value != Py_None) {
+        if (!PyBytes_Check(value)) {
+            PyErr_Format(PyExc_TypeError, "'time_units' is not bytes");
+            goto out;
+        }
+        err = PyBytes_AsStringAndSize(value, &time_units, &time_units_length);
```
Comment (jeromekelleher) [inline python/lwt_interface/tskit_lwt_interface.h:1348]: Likewise here, I think unicode will be a lot simpler for interchange.

File: python/tskit/tables.py [hunk 2]
```diff
@@ -2637,6 +2638,17 @@ def metadata_bytes(self) -> Any:
         """
         return self._ll_tables.metadata
 
+    @property
+    def time_units(self) -> Any:
```
Comment (jeromekelleher) [inline python/tskit/tables.py:2642]: annotation should by ``str``?

**Review by jeromekelleher — COMMENTED** [review]
LGTM, but see point about using unicode in Python-C layer. Can follow the example of the metadata schema, I think.

File: python/_tskitmodule.c [hunk 3]
```diff
@@ -6028,6 +6028,51 @@ TableCollection_get_file_uuid(TableCollection *self, void *closure)
...
+
+    if (TableCollection_check_state(self) != 0) {
+        goto out;
+    }
+    if (arg == NULL) {
+        PyErr_Format(PyExc_AttributeError,
+            "Cannot del time_units, set to empty string (b\"\") to clear.");
+        goto out;
+    }
+    err = PyBytes_AsStringAndSize(arg, &time_units, &time_units_length);
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/_tskitmodule.c]: Yes - good call.

**Comment by benjeffery** [comment]
All fixed up, should be good to go.

File: python/lwt_interface/dict_encoding_testlib.py [hunk 4]
```diff
@@ -202,6 +202,15 @@ def test_missing_sequence_length(self, tables):
         with pytest.raises(TypeError):
             lwt.fromdict(d)
 
+    def test_missing_time_units(self, tables):
+        assert tables.time_units != b""
```
Comment (jeromekelleher) [inline python/lwt_interface/dict_encoding_testlib.py:206]: Probably don't want to check bytes now?

File: python/lwt_interface/tskit_lwt_interface.h [hunk 5]
```diff
@@ -1334,6 +1335,27 @@ parse_table_collection_dict(tsk_table_collection_t *tables, PyObject *tables_dic
...
+    if (value == NULL) {
+        goto out;
+    }
+    if (value != Py_None) {
+        if (!PyUnicode_Check(value)) {
+            PyErr_Format(PyExc_TypeError, "'time_units' is not a string");
+            goto out;
+        }
+        time_units = parse_unicode_arg(value, &time_units_length);
+        if (time_units == NULL) {
```
Comment (jeromekelleher) [inline python/lwt_interface/tskit_lwt_interface.h:1349]: We should try to cover this failure if we can - maybe pass in a non-utf8 encodeable string? Might not be that easy to do though, I guess.

**Review by jeromekelleher — APPROVED** [review]
LGTM - nice to push the coverage up a tiny bit if we could, but no biggie.

File: python/lwt_interface/tskit_lwt_interface.h (hunk 5 above)
Comment (benjeffery) [reply python/lwt_interface/tskit_lwt_interface.h:1349]: Yep extra test with `"\ud861\udd37"` added.

**Pull request merged**